yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,412 @@
+        target: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """Decodes a sequence given the encoded input.
+
+        Decodes until all sequences in a batch have reached [EOS] up to
+        length of `target` args.
+
+        Args:
+                encoder_out (torch.Tensor): batch of encoded input symbols.
+                    Shape: (batch_size x src_len x encoder_hidden*num_dir).
+                encoder_mask (torch.Tensor): mask for the batch of encoded
+                    input symbols. Shape: (batch_size x src_len).
+                target (torch.Tensor): target symbols. Decodes up to
+                    `len(target)` symbols.
+
+        Returns:
+                all_log_probs: (torch.Tensor): tensor of emission
+                    probabilities for each transition state (target symb).
+                    Shape: (tgt_len x batch_size x src_len x vocab_size)
+                all_alignements: (torch.Tensor): tensor of transition
+                    probabilities for each transition state (target symb).
+                    Shape: (tgt_len x batch_size x src_len x src_len)
+        """
+        log_probs, alignment_matrix, decoder_hiddens = self.init_decoding(
+            encoder_out, encoder_mask
+        )
+        all_log_probs, all_alignments = [log_probs], [alignment_matrix]
+        for t in range(target.size(1)):
+            tgt_symbol = target[:, t]
+            log_probs, alignment_matrix, decoder_hiddens = self.decode_step(
+                tgt_symbol.unsqueeze(-1),
+                decoder_hiddens,
+                encoder_out,
+                encoder_mask,
+            )
+            all_log_probs.append(log_probs), all_alignments.append(
+                alignment_matrix
+            )
+        return torch.stack(all_log_probs), torch.stack(all_alignments)
+
+    def decode_step(
+        self, tgt_symbol, decoder_hiddens, encoder_out, encoder_mask
+    ) -> Union[torch.Tensor, torch.Tensor]:
+        """Performs a single decoding step for current state of decoder.
+
+        Args:
+                tgt_symbol (torch.Tensor): tgt symbol for current state.
+                encoder_out (torch.Tensor): batch of encoded input symbols.
+                    Shape: (batch_size x src_len x encoder_hidden*num_dir).
+                encoder_mask (torch.Tensor): mask for the batch of encoded
+                        input symbols of shape (batch_size x src_len).
+
+        Returns:
+                all_log_probs: (torch.Tensor): tensor of emission
+                    probabilities for each transition state (target symb).
+                all_alignements: (torch.Tensor): tensor of transition
+                    probabilities for each transition state (target symb).
+        """
+        decoded = self.decoder(
+            tgt_symbol,
+            decoder_hiddens,
+            encoder_out,
+            encoder_mask,
+        )
+        output, alignment_matrix, decoder_hiddens = (
```

Comment:
Why are we calling the embeddings the alignment matrix?

Reply:
yeah transition probs works better.